### PR TITLE
Fix overlapping x-tick labels on static viz timeseries charts

### PR DESCRIPTION
**Pull request by WiNloSt**
Related to https://github.com/metabase/metabase/issues/24759

[Original commit that introduced the fix](https://github.com/metabase/metabase/pull/25145/commits/6d174ec0a4b9444546a8cc1b43ff3d895c9f4a57) (but not covering all cases apparently).

### Description

The current logic somehow still doesn't account to a situation when the ticks return from `getTicks` or our fallback ticks is vastly different from the `numTicks` passed.

This fix adds an additional check to fix the ticks to be within `numTicks`

### How to verify

1. Create a new question Orders summarize Sum of Total by Create At: Month Filter by Sep 1, 2022 - Aug 31, 2024
    image
2. Add that question to a dashboard
3. Set up a subscription and click *Send email now*

#### After
localhost_email_w2gBCfLM_html

#### Before
localhost_email_Pgdo5CS6_html

### Checklist

- [ ] ~Tests have been added/updated to cover changes in this PR~ Not sure how to test static viz.

### CODE REVIEW

**Review by JesseSDevaney — CHANGES_REQUESTED** [review]
Hi @WiNloSt - thanks for noticing this issue and getting started on the fix :heart: 

This implementation gets most of the way there, but there are a number of required and non-required changes before being complete. I address them individually in the suggested changes comment below.

---
**To-Do**
- [x] Remove `ticks.length <= numTicks` from filter expression
- [x] Remove `numTicks == null` from filter expression
- [x] Debug and find out why there are still more ticks than `numTicks` being plotted
- [x] Create and implement a solution to the underlying issue

File: frontend/src/metabase/static-viz/components/XYChart/utils/ticks.ts [hunk 1]
```diff
@@ -181,7 +181,13 @@ export function fixTimeseriesTicksExceedXTickCount(
         }
       });
 
-    return minLengthTicks;
+    if (minLengthTicks.length <= numTicks) {
+      return minLengthTicks;
+    }
+
+    return minLengthTicks.filter(
+      (_, index, ticks) => index % Math.ceil(ticks.length / numTicks) === 0,
```
Comment (WiNloSt) [inline frontend/src/metabase/static-viz/components/XYChart/utils/ticks.ts:189]: @JesseSDevaney I applied your suggestion + ensured the result ticks length doesn't go over `numTicks`

The original logic was introduced in https://github.com/airbnb/visx/pull/752, there was no explanation why this logic was like that.

I used `Math.ceil` instead to ensure that we don't go over `numTicks` i.e. if we want to chop up 5-length ticks into 4-length ticks we must filter the ticks every other ticks. Because 5/4 = 1.25 if we use `Math.round` we'll still get 1 and the original logic used `ticks.length - 1` divided by `numTicks`, I'm not totally sure why they specifically use `ticks.length - 1` instead of just `ticks.length` since that compares with the use of `Math.round` could result in the ticks length that is longer `numTicks` in many cases.

Considering the static viz which has limited space we'd only see a handful of ticks anyway, so the newly modified logic is accounted for that.

Comment (WiNloSt) [reply frontend/src/metabase/static-viz/components/XYChart/utils/ticks.ts:189]: This, however, changes some of the charts we're rendering e.g. https://percy.io/6a8dd5c9/metabase/builds/30026586/changed/1662698150?browser=chrome&browser_ids=48&subcategories=unreviewed%2Cchanges_requested&viewLayout=overlay&viewMode=new&width=1280&widths=1280

https://github.com/metabase/metabase/blob/645927ceebb0394cda4c077af07ec89c760b8a61/frontend/src/metabase/static-viz/components/XYChart/XYChart.tsx#L150

Since we hard code the `numTicks` to 4 and the previous version of the code doesn't ensure the number of ticks doesn't go over 4 we could sometimes render more than 4 ticks. With the new logic we ensure that we will not render more than 4 ticks and in some cases would make the chart look a little bit worse (see the Percy snapshot above) e.g. when we used to render 5 ticks and now only down to 3

Please let me know if we should still rework this more. I however think this is still better than having a ton of charts having overlapping x tick labels.

File: frontend/src/metabase/static-viz/components/XYChart/utils/ticks.ts [hunk 2]
```diff
@@ -181,7 +181,13 @@ export function fixTimeseriesTicksExceedXTickCount(
...
+    );
```
Comment (JesseSDevaney) [inline frontend/src/metabase/static-viz/components/XYChart/utils/ticks.ts:190]: I did some thinking - if we had more control over the tick creation process, the idealistic solution would look like this

```suggestion
if (numTicks === 1) {
  return [minLengthTicks[0]];
}

return minLengthTicks.reduce((acc, currVal, currIndex, arr) => {
    if (
      acc.length < numTicks &&
      currIndex % (Math.floor((arr.length - numTicks) / (numTicks - 1)) + 1) ===
        0
    ) {
      return [...acc, currVal];
    }

    return acc;
  }, []);
```

But this is only a 'patch' to the actual problem. The real issue is that visx has a bug in the package we are consuming - `getTicks(..., numTicks)` should only be returning a tick array with `<= numTicks` in it, but it is not (so that is we why are having to apply extra logic to filter it out). In my opinion, this leads to poor tick choices because we are only able to select from a pre-selected list of ticks which has different spacing than the optimal choice.

---
**TLDR** (too long did not read)
The issue here is that visx has a bug in the library we are using. So this PR is really only a patch to that problem - it is not fixing the 'real' problem. But it is better to not have 10 overlapping ticks, so this will have to do for now. I will create add this to the over-arching visualization bugs and inconsistencies issue:
- https://github.com/metabase/metabase/issues/33626

**Review by JesseSDevaney — CHANGES_REQUESTED** [review]
Thanks you for updating the logic @WiNloSt! :heart: 

 I did some more digging and my preliminary findings are that this issue is actually related to a bug in the visx library we use. I think the idealistic fix to this requires more effort, so I will be approving this after one small change and then noting the issue in an over-arching visualization bugs and inconsistencies tracker:
- https://github.com/metabase/metabase/issues/33626

---
**Changes**

The only change I want before approving is adding `numTicks == null ||` to the below if expression (since we removed it from the filter expression)
` if (minLengthTicks.length <= numTicks) {`

The other change I mention (in the separate comment) is part of the idealistic one, but we should not implement that here. I will add it to the doc I mentioned above and then get it prioritized as product sees fit.

File: frontend/src/metabase/static-viz/components/XYChart/utils/ticks.ts (hunk 2 above)
Comment (WiNloSt) [reply frontend/src/metabase/static-viz/components/XYChart/utils/ticks.ts:190]: Totally agree with your assessment. We shouldn't have to add too much complexity here, since this is a bug from visx.

File: frontend/src/metabase/static-viz/components/XYChart/utils/ticks.ts [hunk 3]
```diff
@@ -181,7 +181,13 @@ export function fixTimeseriesTicksExceedXTickCount(
         }
       });
 
-    return minLengthTicks;
+    if (minLengthTicks.length <= numTicks) {
```
[the remark this answers is not in the document]
Comment (WiNloSt) [reply frontend/src/metabase/static-viz/components/XYChart/utils/ticks.ts]: Fixed in https://github.com/metabase/metabase/pull/33707/commits/fa0fb626553bde91d1bd990001c36fc048e99aaa

**Comment by WiNloSt** [comment]
Hi @JesseSDevaney. Thanks again for the prompt review, and sorry for the late update, as I just recovered from COVID 🦠.

I've updated the small detail as you suggested.

**Review by JesseSDevaney — APPROVED** [review]

**Comment by JesseSDevaney** [comment]
Can we add a code comment just above the new logic we added to document that it was added because of the visx bug? It would be helpful for future reads of this code.

File: frontend/src/metabase/static-viz/components/XYChart/utils/ticks.ts [hunk 4]
```diff
@@ -159,6 +159,14 @@ export const getYTickWidths = (
   };
 };
 
+/**
```
Comment (WiNloSt) [inline frontend/src/metabase/static-viz/components/XYChart/utils/ticks.ts:162]: @JesseSDevaney I added the comment to the whole function instead, since this is the reason this function exists in the first place, not just the new logic we just added in this PR.
